pkg/catalogv2/git/git.go

```diff
@@ -1,397 +1,480 @@
-		return fmt.Errorf("fetch failure: %w", err)
+func (g *git) fetchAndReset(rev string) error {
+	if err := g.git("-C", g.Directory, "fetch", "origin", "--", rev); err != nil {
+		return err
 	}
+	return g.reset("FETCH_HEAD")
+}
 
-	return nil
+func (g *git) reset(rev string) error {
+	return g.git("-C", g.Directory, "reset", "--hard", rev)
+}
+
+func (g *git) currentCommit() (string, error) {
+	return g.gitOutput("-C", g.Directory, "rev-parse", "HEAD")
 }
 
-// hardReset performs a hard reset of the git repository to a specific commit.
-func (r *Repository) hardReset(reference string) error {
-	var err error
-	resetOpts := r.resetOpts
+func (g *git) gitCmd(output io.Writer, args ...string) error {
+	kv := fmt.Sprintf("credential.helper=%s", `/bin/sh -c 'echo "password=$GIT_PASSWORD"'`)
+	cmd := exec.Command("git", append([]string{"-c", kv}, args...)...)
+	cmd.Env = append(os.Environ(), fmt.Sprintf("GIT_PASSWORD=%s", g.password))
+	stderrBuf := &bytes.Buffer{}
+	cmd.Stderr = stderrBuf
+	cmd.Stdout = output
 
-	switch {
-	case isLocalBranch(reference):
-		branchRef, err := r.repoGogit.Reference(plumbing.ReferenceName(reference), true)
+	if g.agent != nil {
+		c, err := g.injectAgent(cmd)
 		if err != nil {
-			return fmt.Errorf("hardReset failure, branch does not exist locally: %w", err)
+			return err
 		}
-		resetOpts.Commit = branchRef.Hash()
-	case reference == plumbing.HEAD.String():
-		commitHash, err := r.getCurrentCommit()
+		defer c.Close()
+	}
+
+	if len(g.knownHosts) != 0 {
+		f, err := ioutil.TempFile("", "known_hosts")
 		if err != nil {
-			return fmt.Errorf("hardReset failure to get current commit: %w", err)
+			return err
 		}
-		resetOpts.Commit = commitHash
-	default:
-		branchRef, err := r.repoGogit.Reference(plumbing.NewRemoteReferenceName("origin", reference), false)
-		if err != nil {
-			return fmt.Errorf("hardReset failure to get branch reference: %w", err)
+		defer os.RemoveAll(f.Name())
+		defer f.Close()
+
+		if _, err := f.Write(g.knownHosts); err != nil {
+			return err
 		}
-		resetOpts.Commit = branchRef.Hash()
+		if err := f.Close(); err != nil {
+			return fmt.Errorf("closing knownHosts file %s: %w", f.Name(), err)
+		}
+
+		cmd.Env = append(cmd.Env, "GIT_SSH_COMMAND="+fmt.Sprintf("ssh -o UserKnownHostsFile=%s", f.Name()))
+	} else {
+		cmd.Env = append(cmd.Env, "GIT_SSH_COMMAND="+"ssh -o StrictHostKeyChecking=accept-new")
 	}
+	cmd.Env = append(cmd.Env, "GIT_TERMINAL_PROMPT=0")
 
-	// Validate hashCommit and reset options
-	err = resetOpts.Validate(r.repoGogit)
-	if err != nil {
-		return fmt.Errorf("hardReset validation failure: %w", err)
+	if g.insecureTLSVerify {
+		cmd.Env = append(cmd.Env, "GIT_SSL_NO_VERIFY=false")
 	}
 
-	// Open new worktree
-	wt, err := r.repoGogit.Worktree()
-	if err != nil {
-		return fmt.Errorf("hardReset failure on WorkTree: %w", err)
+	if len(g.caBundle) > 0 {
+		f, err := ioutil.TempFile("", "ca-pem-")
```

Comment:
ioutil.TempFile is deprecated: As of Go 1.17, this function simply calls os.CreateTemp.deprecated(default)